Kaggle provided 2 data sets on this project, let's explore these two data sets

In [2]:
# import packages
import pandas as pd
import numpy as np

In [10]:
# load data
df_2018 = pd.read_csv('ks-projects-201801.csv')
df_2016 = pd.read_csv('ks-projects-201612.csv', encoding = 'Windows-1252')

C:\Users\Dell\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:2698: DtypeWarning: Columns (13,14,15) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [11]:
# preview of first 5 rows
df_2018.head()

ID                                               name  \
0  1000002330                    The Songs of Adelaide & Abullah   
1  1000003930      Greeting From Earth: ZGAC Arts Capsule For ET   
2  1000004038                                     Where is Hank?   
3  1000007540  ToshiCapital Rekordz Needs Help to Complete Album   
4  1000011046  Community Film Project: The Art of Neighborhoo...   

         category main_category currency    deadline     goal  \
0          Poetry    Publishing      GBP  2015-10-09   1000.0   
1  Narrative Film  Film & Video      USD  2017-11-01  30000.0   
2  Narrative Film  Film & Video      USD  2013-02-26  45000.0   
3           Music         Music      USD  2012-04-16   5000.0   
4    Film & Video  Film & Video      USD  2015-08-29  19500.0   

              launched  pledged     state  backers country  usd pledged  \
0  2015-08-11 12:12:28      0.0    failed        0      GB          0.0   
1  2017-09-02 04:43:57   2421.0    failed       15      US        100.0   
2  2013-01-12 00:20:50    220.0    failed        3      US        220.0   
3  2012-03-17 03:24:11      1.0    failed        1      US          1.0   
4  2015-07-04 08:35:03   1283.0  canceled       14      US       1283.0   

   usd_pledged_real  usd_goal_real  
0               0.0        1533.95  
1            2421.0       30000.00  
2             220.0       45000.00  
3               1.0        5000.00  
4            1283.0       19500.00

In [12]:
df_2016.head()

ID                                               name   \
0  1000002330                    The Songs of Adelaide & Abullah   
1  1000004038                                     Where is Hank?   
2  1000007540  ToshiCapital Rekordz Needs Help to Complete Album   
3  1000011046  Community Film Project: The Art of Neighborhoo...   
4  1000014025                               Monarch Espresso Bar   

        category  main_category  currency             deadline   goal   \
0          Poetry     Publishing       GBP  2015-10-09 11:36:00   1000   
1  Narrative Film   Film & Video       USD  2013-02-26 00:20:50  45000   
2           Music          Music       USD  2012-04-16 04:24:11   5000   
3    Film & Video   Film & Video       USD  2015-08-29 01:00:00  19500   
4     Restaurants           Food       USD  2016-04-01 13:38:27  50000   

             launched  pledged       state  backers  country  usd pledged   \
0  2015-08-11 12:12:28        0      failed        0       GB            0   
1  2013-01-12 00:20:50      220      failed        3       US          220   
2  2012-03-17 03:24:11        1      failed        1       US            1   
3  2015-07-04 08:35:03     1283    canceled       14       US         1283   
4  2016-02-26 13:38:27    52375  successful      224       US        52375   

  Unnamed: 13 Unnamed: 14 Unnamed: 15  Unnamed: 16  
0         NaN         NaN         NaN          NaN  
1         NaN         NaN         NaN          NaN  
2         NaN         NaN         NaN          NaN  
3         NaN         NaN         NaN          NaN  
4         NaN         NaN         NaN          NaN

In [13]:
# shape & info of data
print(df_2018.shape)
print(df_2018.info())

(378661, 15)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 378661 entries, 0 to 378660
Data columns (total 15 columns):
ID                  378661 non-null int64
name                378657 non-null object
category            378661 non-null object
main_category       378661 non-null object
currency            378661 non-null object
deadline            378661 non-null object
goal                378661 non-null float64
launched            378661 non-null object
pledged             378661 non-null float64
state               378661 non-null object
backers             378661 non-null int64
country             378661 non-null object
usd pledged         374864 non-null float64
usd_pledged_real    378661 non-null float64
usd_goal_real       378661 non-null float64
dtypes: float64(5), int64(2), object(8)
memory usage: 43.3+ MB
None


In [14]:
print(df_2016.shape)
print(df_2016.info())

(323750, 17)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 323750 entries, 0 to 323749
Data columns (total 17 columns):
ID                323750 non-null int64
name              323746 non-null object
category          323745 non-null object
main_category     323750 non-null object
currency          323750 non-null object
deadline          323750 non-null object
goal              323750 non-null object
launched          323750 non-null object
pledged           323750 non-null object
state             323750 non-null object
backers           323750 non-null object
country           323750 non-null object
usd pledged       319960 non-null object
Unnamed: 13       625 non-null object
Unnamed: 14       12 non-null object
Unnamed: 15       4 non-null object
Unnamed: 16       1 non-null float64
dtypes: float64(1), int64(1), object(15)
memory usage: 42.0+ MB
None


#### It seems like the 2018 data contain for projects than the 2016 data, 370,000+ vs. 320,000+

In [15]:
# count of unique values for each variables  
df_2018.nunique()

ID                  378661
name                375764
category               159
main_category           15
currency                14
deadline              3164
goal                  8353
launched            378089
pledged              62130
state                    6
backers               3963
country                 23
usd pledged          95455
usd_pledged_real    106065
usd_goal_real        50339
dtype: int64

In [16]:
df_2016.nunique()

ID                323750
name              321612
category             770
main_category        120
currency              37
deadline          294813
goal                8188
launched          322798
pledged            55598
state                410
backers             3592
country              162
usd pledged        94376
Unnamed: 13          436
Unnamed: 14            8
Unnamed: 15            4
Unnamed: 16            1
dtype: int64

#### There appears to be some obvious discrepancy in terms of numbers of unique values across two data frames (e.g. category, 159 vs 770)

This equires further exploration

In [19]:
# There is issue accessing columns from df_1016 due to white space 
# strip the white-spaces from df_2016's column names
print(df_2016.columns.tolist())
df_2016.columns = df_2016.columns.str.strip()
print(df_2016.columns.tolist())

['ID', 'name', 'category', 'main_category', 'currency', 'deadline', 'goal', 'launched', 'pledged', 'state', 'backers', 'country', 'usd pledged', 'Unnamed: 13', 'Unnamed: 14', 'Unnamed: 15', 'Unnamed: 16']
['ID', 'name', 'category', 'main_category', 'currency', 'deadline', 'goal', 'launched', 'pledged', 'state', 'backers', 'country', 'usd pledged', 'Unnamed: 13', 'Unnamed: 14', 'Unnamed: 15', 'Unnamed: 16']


### Let's take a look at some differences between the two data frames with a bit more details
Length

In [23]:
# length of the two data frames
print('length df_2016: ', len(df_2016))
print('length df_2018: ', len(df_2018))
# difference in length 
diff_len = len(df_2018) - len(df_2016)
print('length difference: ',diff_len)

length df_2016:  323750
length df_2018:  378661
length difference:  54911


check for duplicated projects within each data frame

In [24]:
print(sum(df_2016['ID'].duplicated()))
print(sum(df_2018['ID'].duplicated()))

0
0


#### There seems to be no duplicated projects within each data frames
Now, check for duplicated projects between data frames

In [30]:
# now, outer join with concatenation vertically & check if there is duplicate across two data frames
df_concat = pd.concat([df_2016, df_2018], join = 'outer', axis = 0)
# check length
print('length of concatenated df: ',len(df_try))
# check duplicates in 'ID'
duplicate_count = sum(df_concat['ID'].duplicated())
print('number of duplicated projects between two data frames: ', duplicate_count)

C:\Users\Dell\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  


length of concatenated df:  702411
number of duplicated projects between two data frames:  323750


It can be a reasonable suspectation that all 2016 projects are duplicated in the 2018 data, let's test this assumption

In [28]:
# check if the amount of duplicate equals the amount of entries in df_2016
print(duplicate_count == len(df_2016))

True


In [31]:
# are all these entries indeed from df_2016? 

# df_try_dupID - the duplicated 'ID' in df_try
df_concat_dupID = df_try[df_try['ID'].duplicated()].sort_values(['ID'])['ID'].reset_index(drop = True)

# df_2016_ID 
df_2016_ID = df_2016.sort_values(['ID'])['ID'].reset_index(drop = True)

# are they the same?
print(sum(df_concat_dupID != df_2016_ID))

0


Indeed, all entries with duplicated 'ID' appear in both 2016 & 2018, 2018 is the updated version of 2016 

#### Now, let's inspect the discrepancy in terms of numbers of unique values across two data frames (e.g. category, 159 vs 770)

In [33]:
# let's look at the unique values of each columns in both data frames
print(df_2018.nunique(), '\n\n')
print(df_2016.nunique())

ID                  378661
name                375764
category               159
main_category           15
currency                14
deadline              3164
goal                  8353
launched            378089
pledged              62130
state                    6
backers               3963
country                 23
usd pledged          95455
usd_pledged_real    106065
usd_goal_real        50339
dtype: int64 


ID               323750
name             321612
category            770
main_category       120
currency             37
deadline         294813
goal               8188
launched         322798
pledged           55598
state               410
backers            3592
country             162
usd pledged       94376
Unnamed: 13         436
Unnamed: 14           8
Unnamed: 15           4
Unnamed: 16           1
dtype: int64


Other than category, there is an obvious discrepancy on "main_category" accross the data frames as well, we will inspect this first

In [35]:
# subsets and merge to form a new data frames to compare "main_category" across data frames
df_compare = df_2018[["ID", "main_category"]].merge(df_2016[["ID", "main_category"]], how="left", on='ID', 
                                                    suffixes=("_2018", "_2016"))

In [36]:
# head rows of the new data frames
df_compare.head()

ID main_category_2018 main_category_2016
0  1000002330         Publishing         Publishing
1  1000003930       Film & Video                NaN
2  1000004038       Film & Video       Film & Video
3  1000007540              Music              Music
4  1000011046       Film & Video       Film & Video

In [44]:
# let's subset the entries where there are different "main_category" in different years
df_compare[(df_compare['main_category_2018'] != df_compare['main_category_2016']) & (df_compare['main_category_2016'].notnull())] 

ID main_category_2018       main_category_2016
1693    1008705746              Music                    Metal
1819    1009317190               Food                Cookbooks
2243    1011687764         Technology                      Web
2814     101453314       Film & Video                   Shorts
3577    1018515072            Theater                    Plays
5656    1028998661         Technology                 Hardware
5683     102911383              Games            Playing Cards
5895    1030235554              Music              World Music
7667    1038703212              Games             Mobile Games
8158    1041248261         Technology         Camera Equipment
8903    1044922583              Music          Classical Music
9140    1046161804                Art           Conceptual Art
11644   1058376015         Publishing               Nonfiction
11835   1059393119             Design           Product Design
13074   1065632207       Film & Video              Documentary
13174     10661684       Film & Video              Documentary
14025   1070559418       Film & Video                   Shorts
14128   1071067313              Games              Video Games
16032   1081139420              Music   50 Years in the Making
16237   1082145521       Film & Video              Documentary
17019   1086338488              Music           Country & Folk
17455    108863281                Art              Mixed Media
17767   1090187010             Comics              Comic Books
18441   1093662806             Design           Product Design
19150   1097082409             Crafts        Retro Gaming art.
19978   1101406194              Music          Classical Music
20611   1104693057        Photography                   Places
21056   1106734072               Food                   Events
21539   1109234649         Publishing                  Fiction
21830   1110576549         Publishing               Nonfiction
...            ...                ...                      ...
353002   868352965       Film & Video              Documentary
353646   871484997         Publishing                Art Books
353713   871787114         Publishing              Periodicals
357068   888963653              Games              Video Games
357323   890269949            Fashion            Restore Pride
358136   894430743       Film & Video           Narrative Film
358838   898132486              Games              Video Games
360150   904800959         Technology                 Hardware
362688   917983232              Music                     Rock
365025   929880656            Fashion                  Apparel
365107   930244233         Publishing                  Fiction
365132   930337808       Film & Video           Narrative Film
367467   942417221              Music          Classical Music
367540   942697158              Music                     Rock
367900   944429286                Art              Mixed Media
368065    94521031       Film & Video                   Shorts
368181    94573368         Publishing                  Fiction
368347   946554330         Publishing               Nonfiction
368719   948525980              Games              Video Games
369548   952768436              Music                  Hip-Hop
369684   953402698                Art                Sculpture
370244   956497576         Technology        Space Exploration
370786   959304168                Art              Mixed Media
373535    97337924              Music           Country & Folk
375472   983555190              Music                     Kids
375787   985310209              Games           Tabletop Games
375849   985600912                Art             Illustration
376526   989007729       Film & Video           Narrative Film
376780   990511774              Music               Indie Rock
377943   996542939       Film & Video                   Shorts

[444 rows x 3 columns]

#### We can see the general trend of a more specific 2016 category being placed under a broader 2018 category, thus the discrepancy is caused by update of data, not inherant differences between data frames

We are going to assume the same for the discrepancy in other columns

In [45]:
# let's see how many projects in these subsets are included because the the "main_category" from 2016 are missing
NA_2016 = len(df_compare[(df_compare['main_category_2018'] != df_compare['main_category_2016']) & (df_compare['main_category_2016'].isnull())])
print('Out of this subset, ', NA_2016, ' projects from 2016 are missing the information on "main_category"')

out of this subset,  54911  projects from 2016 are missing the information on "main_category"


In [48]:
# let's see how many projects in these subsets are included because the the "main_category" from 2016 are missing
NA_2018 = len(df_compare[(df_compare['main_category_2018'] != df_compare['main_category_2016']) & (df_compare['main_category_2018'].isnull())])
print('Out of this subset, ', NA_2018, ' projects from 2018 are missing the information on "main_category"')

Out of this subset,  0  projects from 2018 are missing the information on "main_category"


#### Which means that, the 2018 data does not only update categorical values from "specific and many" to "broad and few", it also updates a lot of missing information

### Use 2018 data because it is updated and more comple